### PR TITLE
Fix fake memory leaks in some test cases [databricks]

### PR DESCRIPTION
**Pull request by res-life**
Contributes to #5854

### Problem
Prints RapidsHostMemoryStore.pool leaked error log when running Rapids Accelerator test cases.
```
All tests passed.

22/06/27 17:45:57.298 Thread-7 ERROR HostMemoryBuffer: A HOST BUFFER WAS LEAKED (ID: 1 7f8557fff010)
22/06/27 17:45:57.303 Thread-7 ERROR MemoryCleaner: Leaked host buffer (ID: 1): 2022-06-27 09:45:16.0171 UTC: INC
java.lang.Thread.getStackTrace(Thread.java:1559)
ai.rapids.cudf.MemoryCleaner$RefCountDebugItem.<init>(MemoryCleaner.java:301)
ai.rapids.cudf.MemoryCleaner$Cleaner.addRef(MemoryCleaner.java:82)
ai.rapids.cudf.MemoryBuffer.incRefCount(MemoryBuffer.java:232)
ai.rapids.cudf.MemoryBuffer.<init>(MemoryBuffer.java:98)
ai.rapids.cudf.HostMemoryBuffer.<init>(HostMemoryBuffer.java:196)
ai.rapids.cudf.HostMemoryBuffer.<init>(HostMemoryBuffer.java:192)
ai.rapids.cudf.HostMemoryBuffer.allocate(HostMemoryBuffer.java:144)
com.nvidia.spark.rapids.RapidsHostMemoryStore.<init>(RapidsHostMemoryStore.scala:38)
```

### Root cause
RapidsHostMemoryStore.pool is not closed before MemoryCleaner checking the leaks.
It's actually not a leak, it's caused by hooks execution order.
RapidsHostMemoryStore.pool is closed in the [Spark executor plugin hook](https://github.com/apache/spark/blob/v3.3.0/core/src/main/scala/org/apache/spark/executor/Executor.scala#L351toL381).

plugins.foreach(_.shutdown())  // this line will eventually close the RapidsHostMemoryStore.pool
The close path is:
```
  The close path is: 
    Spark executor plugin hook ->
      RapidsExecutorPlugin.shutdown ->
        GpuDeviceManager.shutdown ->
          RapidsBufferCatalog.close() ->
            RapidsHostMemoryStore.close ->
              RapidsHostMemoryStore.pool.close ->
```

### Solution
First, remove the default hook in MemoryCleaner, then add the default hook by leveraging Spark ShutdownHookManager

See the cuDF side change:  https://github.com/rapidsai/cudf/pull/11161

Signed-off-by: Chong Gao <res_life@163.com>

### CODE REVIEW

**Comment by res-life** [comment]
Depends on https://github.com/rapidsai/cudf/pull/11161

Tested, and it works.

**Review by revans2 — DISMISSED** [review]

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/Plugin.scala [hunk 1]
```diff
@@ -259,6 +262,34 @@ class RapidsExecutorPlugin extends ExecutorPlugin with Logging {
...
+  /**
+   * Re-register leaks checking hook if configured.
+   */
+  private def ReRegisterCheckLeakHook: Unit = {
+    // DEFAULT_SHUTDOWN_THREAD in MemoryCleaner is responsible to check the leaks at shutdown time,
+    // it expects all other hooks are done before the checking
+    // as other hooks will close some resources.
+
+    val REF_COUNT_DEBUG_STR = System.getProperty(MemoryCleaner.REF_COUNT_DEBUG_KEY, "false")
+    if (REF_COUNT_DEBUG_STR.equalsIgnoreCase("true")) {
```
Comment (gerashegalov) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/Plugin.scala:274]: nit:
```suggestion
    if (java.lang.Boolean.getBoolean(MemoryCleaner.REF_COUNT_DEBUG_KEY)) {
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/Plugin.scala [hunk 2]
```diff
@@ -259,6 +262,34 @@ class RapidsExecutorPlugin extends ExecutorPlugin with Logging {
...
+      // See issue: https://github.com/NVIDIA/spark-rapids/issues/5854
+      //
+      // `Spark ShutdownHookManager` leverages `Hadoop ShutdownHookManager` to manage hooks with
+      // priority. The priority parameter will guarantee the execution order.
+      //
+      // Here also use `Hadoop ShutdownHookManager` to add a lower priority hook.
+      // 20 priority is small enough, will run after Spark hooks.
+      // Note: `ShutdownHookManager.get()` is a singleton
+      org.apache.hadoop.util.ShutdownHookManager.get().addShutdownHook(
+        MemoryCleaner.DEFAULT_SHUTDOWN_RUNNABLE, 20)
```
Comment (gerashegalov) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/Plugin.scala:289]: nit: if you follow the suggestion on the cudf PR you could do
```suggestion
      // Shutdown hooks are executed concurrently in JVM, and there is no execution order guarantee.
      // See the doc of `Runtime.addShutdownHook`.
      // Some resources are closed in Spark hooks.
      // Here we should wait Spark hooks to be done, or a false leak will be detected.
      // See issue: https://github.com/NVIDIA/spark-rapids/issues/5854
      //
      // `Spark ShutdownHookManager` leverages `Hadoop ShutdownHookManager` to manage hooks with
      // priority. The priority parameter will guarantee the execution order.
      //
      // Here also use `Hadoop ShutdownHookManager` to add a lower priority hook.
      // 20 priority is small enough, will run after Spark hooks.
      // Note: `ShutdownHookManager.get()` is a singleton
      org.apache.hadoop.util.ShutdownHookManager.get().addShutdownHook(
              MemoryCleaner.removeDefaultShutdownHook(), 20)
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/Plugin.scala [hunk 3]
```diff
@@ -259,6 +262,34 @@ class RapidsExecutorPlugin extends ExecutorPlugin with Logging {
     }
   }
 
+  /**
+   * Re-register leaks checking hook if configured.
+   */
+  private def ReRegisterCheckLeakHook: Unit = {
```
Comment (gerashegalov) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/Plugin.scala:268]: nit : this is a method (should start with a lowercase) and it has side effects, should use empty parameter list in parens. 
```suggestion
  private def reRegisterCheckLeakHook(): Unit = {
```

Comment (res-life) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/Plugin.scala]: Done.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/Plugin.scala (hunk 2 above)
Comment (res-life) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/Plugin.scala]: Done.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/Plugin.scala (hunk 1 above)
Comment (res-life) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/Plugin.scala]: Done.

**Review by gerashegalov — DISMISSED** [review]
LGTM

**Review by gerashegalov — APPROVED** [review]

**Review by revans2 — APPROVED** [review]

**Review by abellina — APPROVED** [review]

**Comment by abellina** [comment]
build

**Comment by res-life** [comment]
build

**Comment by res-life** [comment]
build

**Comment by res-life** [comment]
build

**Pull request merged**